### PR TITLE
Attendance off campus

### PR DESCRIPTION
**Pull request by aaronbrowder**
This is a new feature in the attendance system that allows users to record off campus time, i.e. if a student leaves campus and returns during the day. The user can add this data using a new tab in Attendance called "off campus time". Off campus time does not count towards a student's attendance hours for the day. I wasn't sure if this should be hidden behind a feature flag in the settings, so I didn't create one.

### CODE REVIEW

**Comment by schmave** [comment]
This seems like a very useful feature. Thanks for working on it! 

I'm curious about why you created a new class `OffCampusEvent` that holds data that looks very similar to what is in `AttendanceDay`. Seems like it would simplify the code to use AttendanceDay directly. What do you think?

Why did you decide to have this data be entered on a separate tab instead of within the existing editing interface?

**Comment by aaronbrowder** [comment]
If by "existing editing interface" you mean the editWeek page, there seems to be clearly not enough space on that page. And I didn't want to clutter up a general use page with something that's only going to be used occasionally for a minority of students.

Since off campus data is entered on its own page, I think it makes sense to have a separate view model object for that page (`OffCampusEvent`). This seems to satisfy the single responsibility principle and make the code cleaner.

**Comment by schmave** [comment]
Alright, that makes sense. I'll do some testing and push this on Thursday or Friday probably.

**Comment by aaronbrowder** [comment]
Ok, thanks!

**Comment by schmave** [comment]
Hey Aaron, I checked this out last night. I'd like this to go behind a feature flag in the settings page since it's new and is likely not EXACTLY what everyone wants. Other than that, looks fine to me. 

I think it would be nice to be able to edit this from the regular weekly editing screen, and see the data displayed on the regular weekly view screen as well. I also think it would be nice to add a check box for days with off-campus activity here:

image

But I know that there is only so much time in the day! So with a feature flag I think it's great.

I did a little refactoring to remove OffCampusEvent because I'd rather have fewer lines of code than purpose-built objects, and also made it so that this new code uses the existing time and date formatters (especially Application.formatDateMdy, which supports swapping month and day depending on a organization setting). Github is not letting me push that commit here but I will merge it in when I merge this.

**Comment by aaronbrowder** [comment]
Ok, I added the feature flag.

**Comment by schmave** [comment]
Thanks! Will get this merged soon.

File: app/controllers/Attendance.java [hunk 1]
```diff
@@ -288,12 +287,17 @@ public Result viewPersonReport(Integer person_id, String start_date_str, String
...
+        boolean has_off_campus_time = false;
+
         for (AttendanceDay day : days) {
             if (day.code != null || day.start_time == null || day.end_time == null) {
                 stats.incrementCodeCount(codes_map.get(day.code));
             } else {
                 stats.incrementAttendance(day);
             }
+            if (day.off_campus_departure_time != null || day.off_campus_return_time != null) {
+                has_off_campus_time = true;
```
Comment (schmave) [inline app/controllers/Attendance.java]: When I looked at this last weekend, I didn't notice that you only show the two extra columns on the person attendance page if that person has some off campus data entered. I think that's quite elegant.

**Pull request merged**